### PR TITLE
Trainer

### PR DESCRIPTION
**Pull request by emailweixu**
Still need checkpoint(@Haichao-Zhang) , ActorCriticAlgorithm (@Haichao-Zhang)  environment (@runjerry), metrics(@le-horizon) , and summary (@runjerry) to run.

Remove all PolicyDrivers, (On/Off)PolicyTrainer.
Add train_iter() to RLAlgorithm.
Please take a look to see if the overall structure is fine.

Issue  TTP: OnPolicyDriver #347  TTP: OnPolicyTrainer #346  RLAlgorithm interface #348

### CODE REVIEW

**Comment by le-horizon** [comment]
@emailweixu Branch torch_metrics has metrics.  Still need to fix code style / docstrings, and fix deque.  Otherwise, metrics is already working (passes unit tests).  I'll take a look at your code tmr.

File: alf/trainers/policy_trainer.py [hunk 1]
```diff
@@ -387,49 +333,67 @@ def _markdownify(paragraph):
...
+        global_step = alf.summary.get_global_counter()
+        checkpointer = Checkpointer(
             ckpt_dir=os.path.join(self._train_dir, 'algorithm'),
-            max_to_keep=self._checkpoint_max_to_keep,
             algorithm=self._algorithm,
             metrics=metric_utils.MetricsGroup(self._driver.get_metrics(),
                                               'metrics'),
             global_step=global_step)
-        checkpointer.initialize_or_restore()
+        checkpointer.load()
```
Comment (Haichao-Zhang) [inline alf/trainers/policy_trainer.py]: Do we need to update the ```global_step``` using the value recovered from the checkpoint here?

File: alf/trainers/policy_trainer.py [hunk 2]
```diff
@@ -387,49 +333,67 @@ def _markdownify(paragraph):
...
+        time_step.is_first())
+    transformed_time_step = algorithm.transform_timestep(time_step)
+    policy_step = algorithm.predict(transformed_time_step, policy_state,
+                                    epsilon_greedy)
+    next_time_step = env.step(policy_step.output)
+
+    exp = alf.data_structures.make_experience(time_step, policy_step,
+                                              policy_state)
+    for metric in metrics:
+        metric.observe(exp)
```
Comment (le-horizon) [inline alf/trainers/policy_trainer.py]: metric(time_step) is the API defined here:
https://github.com/HorizonRobotics/alf/blob/torch_metrics/alf/metrics/metrics.py#L40

File: alf/algorithms/rl_algorithm.py [hunk 3]
```diff
@@ -402,7 +398,7 @@ def predict(self, time_step: ActionTimeStep, state, epsilon_greedy):
     PREPARE_SPEC = 3
 
     @abstractmethod
-    def rollout(self, time_step: ActionTimeStep, state, mode):
+    def rollout(self, time_step: TimeStep, state, mode):
```
Comment (hnyu) [inline alf/algorithms/rl_algorithm.py]: Maybe now `rollout` can be better called `predict_step_for_training`? Not to confuse it with `unroll`

File: alf/algorithms/rl_algorithm.py [hunk 4]
```diff
@@ -363,16 +364,11 @@ def summarize_metrics(self):
...
-            name='memory_usage')
-        if not tf.executing_eagerly():
-            mem.set_shape(())
-        tf.summary.scalar(name='memory_usage', data=mem)
+        mem = self._proc.memory_info().rss // 1e6
+        alf.summary.scalar(name='memory_usage', data=mem)
 
     # Subclass may override predict() to allow more efficient implementation
-    def predict(self, time_step: ActionTimeStep, state, epsilon_greedy):
+    def predict(self, time_step: TimeStep, state, epsilon_greedy):
```
Comment (hnyu) [inline alf/algorithms/rl_algorithm.py]: Rename `predict` to `predict_step`?

File: alf/algorithms/rl_algorithm.py [hunk 5]
```diff
@@ -538,3 +528,74 @@ def calc_loss(self, training_info: TrainingInfo):
             the tensors in loss_info should be (T, B)
         """
         pass
+
+    def unroll(self, time_step, policy_state, unroll_length):
```
Comment (hnyu) [inline alf/algorithms/rl_algorithm.py]: Comment?

File: alf/utils/common.py [hunk 6]
```diff
@@ -973,20 +941,11 @@ def decorate(f):
...
-
-        os.environ["TF_DETERMINISTIC_OPS"] = str(1)
-    else:
+    if seed is None:
         seed = os.getpid() + int(time.time())
-
     random.seed(seed)
     np.random.seed(seed)
-    tf.random.set_seed(seed)
+    torch.random.manual_see(seed)
```
Comment (hnyu) [inline alf/utils/common.py]: `torch.manual_seed` See https://pytorch.org/docs/stable/notes/randomness.html

File: alf/algorithms/rl_algorithm.py (hunk 4 above)
Comment (emailweixu) [reply alf/algorithms/rl_algorithm.py]: changed to predict_step, rollout_step

File: alf/algorithms/rl_algorithm.py (hunk 5 above)
Comment (emailweixu) [reply alf/algorithms/rl_algorithm.py]: added

File: alf/utils/common.py (hunk 6 above)
Comment (emailweixu) [reply alf/utils/common.py]: fixed

File: alf/algorithms/on_policy_algorithm.py [hunk 7]
```diff
@@ -13,52 +13,49 @@
...
+from alf.data_structures import TimeStep, Experience, TrainingInfo
 
 
-class OnPolicyAlgorithm(OffPolicyAlgorithm):
+class OnPolicyAlgorithm(RLAlgorithm):
     """
     OnPolicyAlgorithm works with alf.drivers.on_policy_driver.OnPolicyDriver
     to do training at the time of policy rollout.
 
     User needs to implement rollout() and train_complete().
```
Comment (Haichao-Zhang) [inline alf/algorithms/on_policy_algorithm.py]: ```rollout()``` -> ```rollout_step()```

Currently there is a mis-match of the name for this function: in implementation it is ```rollout()```, in documentation it is a mixture of ```rollout()``` and ```rollout_step()```.

File: alf/algorithms/on_policy_algorithm.py [hunk 8]
```diff
@@ -13,52 +13,49 @@
...
     ```python
-    with GradientTape as tape:
-        for _ in range(train_interval):
-            policy_step = rollout(time_step, policy_step.state)
-            action = sample action from policy_step.action
-            collect necessary information and policy_step.info into training_info
-            time_step = env.step(action)
-    train_complete(tape, training_info)
+    for _ in range(unroll_length):
+        policy_step = rollout(time_step, policy_step.state)
```
Comment (Haichao-Zhang) [inline alf/algorithms/on_policy_algorithm.py]: ```rollout()``` -> ```rollout_step()```

File: alf/algorithms/rl_algorithm.py [hunk 9]
```diff
@@ -37,22 +38,33 @@ class RLAlgorithm(Algorithm):
     RLAlgorithm provide basic functions and generic interface for rl algorithms.
 
     The key interface functions are:
-    1. predict(): one step of computation of action for evaluation.
-    2. rollout(): one step of computation for rollout. rollout() is used for
-        collecting experiences during training. Different from `predict`,
-        `rollout` may include addtional computations for training.
+    1. predict_step(): one step of computation of action for evaluation.
+    2. rollout_step(): one step of computation for rollout. rollout() is used for
```
Comment (Haichao-Zhang) [inline alf/algorithms/rl_algorithm.py]: ```rollout()``` -> ```rollout_step()```

File: alf/algorithms/rl_algorithm.py [hunk 10]
```diff
@@ -70,10 +82,13 @@ def __init__(self,
             action_spec (nested BoundedTensorSpec): representing the actions.
             train_state_spec (nested TensorSpec): for the network state of
                 `rollout()`
```
Comment (Haichao-Zhang) [inline alf/algorithms/rl_algorithm.py]: ```rollout()``` -> ```rollout_step()```

File: alf/algorithms/rl_algorithm.py (hunk 3 above)
Comment (Haichao-Zhang) [reply alf/algorithms/rl_algorithm.py]: ```rollout()``` -> ```rollout_step()```

File: alf/algorithms/rl_algorithm.py [hunk 11]
```diff
@@ -420,8 +424,8 @@ def rollout(self, time_step: ActionTimeStep, state, mode):
                     rollout() should not make any side effect during this, such
```
Comment (Haichao-Zhang) [inline alf/algorithms/rl_algorithm.py]: ```rollout()``` -> ```rollout_step()```

File: alf/algorithms/rl_algorithm_test.py [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+            input_size=2, num_actions=3)
+
+    def is_on_policy(self):
+        return True
+
+    def predict(self, time_step: TimeStep, state, epsilon_greedy):
+        dist = self._proj_net(time_step.observation)
+        return AlgStep(output=dist.sample(), state=(), info=())
+
+    def rollout(self, time_step: TimeStep, state, mode):
```
Comment (Haichao-Zhang) [inline alf/algorithms/rl_algorithm_test.py]: ```rollout()``` -> ```rollout_step()```

File: alf/algorithms/algorithm.py [hunk 13]
```diff
@@ -309,10 +292,19 @@ def convert_train_state_to_predict_state(self, state):
...
+        return common.zeros_from_spec(self._rollout_state_spec, batch_size)
+
+    def get_initial_train_step_state(self, batch_size):
+        return common.zeros_from_spec(self._train_state_spec, batch_size)
+
     #------------- User need to implement the following functions -------
 
     # Subclass may override predict() to allow more efficient implementation
-    def predict(self, inputs, state=None):
+    def predict_step(self, inputs, state=None):
```
Comment (Haichao-Zhang) [inline alf/algorithms/algorithm.py]: ```predict``` -> ```predict_step``` in doc

File: alf/algorithms/rl_algorithm.py [hunk 14]
```diff
@@ -70,10 +82,13 @@ def __init__(self,
...
             train_state_spec (nested TensorSpec): for the network state of
                 `rollout()`
-            action_distribution_spec (nested DistributionSpec): for the action
-                distributions.
+            env (Environment): The environment to interact with. env is a batched
+                environment, which means that it runs multiple simulations
+                simultateously.
+            config (TrainerConfig): config for training.
             predict_state_spec (nested TensorSpec): for the network state of
                 `predict()`. If None, it's assumed to be the same as train_state_spec
```
Comment (Haichao-Zhang) [inline alf/algorithms/rl_algorithm.py]: ```predict``` -> ```predict_step```

File: alf/algorithms/rl_algorithm.py (hunk 4 above)
Comment (Haichao-Zhang) [reply alf/algorithms/rl_algorithm.py]: ```predict```  -> ```predict_step```

File: alf/algorithms/rl_algorithm_test.py [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+            debug_summaries=True,
+            name="MyAlg")
+
+        self._proj_net = alf.networks.CategoricalProjectionNetwork(
+            input_size=2, num_actions=3)
+
+    def is_on_policy(self):
+        return True
+
+    def predict(self, time_step: TimeStep, state, epsilon_greedy):
```
Comment (Haichao-Zhang) [inline alf/algorithms/rl_algorithm_test.py]: ```predict``` -> ```predict_step```

File: alf/algorithms/rl_algorithm_test.py [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+    def test_rl_algorithm(self):
+        config = Config()
+        env = MyEnv(batch_size=3)
+        alg = MyAlg(env=env, config=config)
+        for _ in range(100):
+            alg.train_iter()
+
+        time_step = common.get_initial_time_step(env)
+        state = alg.get_initial_predict_state(env.batch_size)
+        policy_step = alg.rollout(time_step, state, mode=RLAlgorithm.ROLLOUT)
```
Comment (Haichao-Zhang) [inline alf/algorithms/rl_algorithm_test.py]: ```rollout``` -> ```rollout_step```

**Comment by emailweixu** [comment]
Now the trainer can run successfully with a dummy environment and is ready for review and merge.
@runjerry you can pull this to your env branch to test your code
@Haichao-Zhang you can test your ac algorithm branch on top of this.

File: alf/utils/common.py [hunk 17]
```diff
@@ -925,20 +894,33 @@ def decorate(f):
...
 
-        os.environ["TF_DETERMINISTIC_OPS"] = str(1)
-    else:
+    Args:
+        seed (int|None): seed to be used. If None, a default seed based on
+            pid and time will be used.
+    Returns:
+        The seed being used if `seed` is None. Note that you may not reproduce
+        a previous experiment even if use use its seed because when seed is None,
+        cudnn.deterministic is not set to True
```
Comment (hnyu) [inline alf/utils/common.py]: I guess the real reason is that we cannot directly set the seed for cudnn given the previous seed?

Comment (hnyu) [reply alf/utils/common.py]: And the comment should be more precise: when someone runs an experiment with a pre-selected manual seed, he can definitely reproduce the results; however, if he runs the experiment with seed=None and later recovers the seed from the tensorboard, and if cudnn is used in the code, then there is no guarantee that the results will be reproduced with the recovered seed.

File: alf/algorithms/rl_algorithm.py [hunk 18]
```diff
@@ -117,11 +136,43 @@ def __init__(self,
...
-            alg.__setattr__(property_name, value)
+        for child in self._get_children():
+            if isinstance(child, RLAlgorithm):
+                child.__setattr__(property_name, value)
+
+    @abstractmethod
+    def is_on_policy(self):
+        """Whehter this algorithm is an on-policy algorithm.
+
+        If it's on-policy algoirhtm, train_iter() will use
```
Comment (Haichao-Zhang) [inline alf/algorithms/rl_algorithm.py]: algoirhtm - > algorithm

File: alf/algorithms/rl_algorithm.py [hunk 19]
```diff
@@ -185,60 +217,42 @@ def experience_spec(self):
...
 
-    def set_metrics(self, metrics=[]):
-        """Set metrics.
+    def get_step_metrics(self):
+        """Get step metrics that used for generating summaries against
 
-        metrics (list[TFStepMetric]): An optional list of metrics
-            len(metrics) >= 2 as required by calling "self._metrics[:2]" in training_summary()
+        Returns:
+             list[TFStepMetric]: step metrics `EnvironmentSteps` and `NumberOfEpisodes`
```
Comment (Haichao-Zhang) [inline alf/algorithms/rl_algorithm.py]: ```TFStepMetric``` -> ```StepMetric```

File: alf/algorithms/rl_algorithm.py [hunk 20]
```diff
@@ -185,60 +217,42 @@ def experience_spec(self):
...
         """
-        self._metrics = metrics
+        return self._metrics[:2]
+
+    def get_metrics(self):
+        """Returns the metrics monitored by this driver.
+
+        Returns:
+            list[TFStepMetric]
```
Comment (Haichao-Zhang) [inline alf/algorithms/rl_algorithm.py]: ```TFStepMetric``` -> ```StepMetric```

File: alf/trainers/policy_trainer.py [hunk 21]
```diff
@@ -443,12 +277,13 @@ def play(root_dir,
         root_dir (str): same as the root_dir used for `train()`
         env (TFEnvironment): the environment
         algorithm (OnPolicyAlgorithm): the training algorithm
-        checkpoint_name (str): name of the checkpoint (e.g. 'ckpt-12800`).
-            If None, the latest checkpoint under train_dir will be used.
+        checkpoint_step (int|str): the number of training steps which is used to
+            specify the checkpoint to be loaded. If glocheckpoint_stepbal_step is 'latest',
```
Comment (Haichao-Zhang) [inline alf/trainers/policy_trainer.py]: ```glocheckpoint_stepbal_step ``` -> ```checkpoint_step```

File: alf/utils/common.py (hunk 17 above)
Comment (emailweixu) [reply alf/utils/common.py]: updated

File: alf/algorithms/rl_algorithm.py (hunk 18 above)
Comment (emailweixu) [reply alf/algorithms/rl_algorithm.py]: fixed

File: alf/algorithms/rl_algorithm.py (hunk 19 above)
Comment (emailweixu) [reply alf/algorithms/rl_algorithm.py]: Fixed

File: alf/algorithms/rl_algorithm.py (hunk 20 above)
Comment (emailweixu) [reply alf/algorithms/rl_algorithm.py]: fixed

File: alf/trainers/policy_trainer.py (hunk 21 above)
Comment (emailweixu) [reply alf/trainers/policy_trainer.py]: fixed
